Add quickstart guide and concepts, issue #108

vfdev-5 opened the pull request:
I propose to address the issue 108 in multiple PRs. Here is the first one with "Quickstart" notes.
Other notes can be added later. 

@alykhantejani @jasonkriss could you review this note, please ?

alykhantejani commented on docs/source/quickstart.rst:
```diff
@@ -1,2 +1,157 @@
+                                            })
+
+    @trainer.on(Events.ITERATION_COMPLETED)
+    def log_training_loss(engine):
```
nit: I would rename this input arg to `trainer` to make it clear its this engine that's passed in

alykhantejani commented on docs/source/quickstart.rst:
```diff
@@ -1,2 +1,157 @@
+        print("Epoch[{}] Loss: {:.2f}".format(engine.state.epoch, len(train_loader), engine.state.output))
+
+    @trainer.on(Events.EPOCH_COMPLETED)
+    def log_validation_results(engine):
```
nit: I would rename this input arg to `trainer` to make it clear its this engine that's passed in

alykhantejani commented on docs/source/quickstart.rst:
```diff
@@ -1,2 +1,157 @@
+Next we define trainer and evaluator engines. The main component of Ignite is the :class:`Engine`, an abstraction over your
+training loop. Getting started with the engine is easy, the constructor only requires one things:
+
+- `update_function`: a function which is passed a batch and passes data through and updates your model
```
It's now also passed the engine too

alykhantejani commented on docs/source/quickstart.rst:
```diff
@@ -1,2 +1,157 @@
+        y_pred = model(x)
+        return to_tensor(y_pred, cpu=not cuda), to_tensor(y, cpu=not cuda)
+
+Remark that the helper function :meth:`create_supervised_evaluator` to create an evaluator accepts an
```
nit:Remark -> Note

alykhantejani commented on docs/source/quickstart.rst:
```diff
@@ -1,2 +1,157 @@
+
+
+The most interesting part of the code snippet is adding event handlers. :class:`Engine` allows to add handlers on
+various events that fired during the run. When event is fired, attached handlers (functions) are executed. Thus, for
```
When **an** event...

alykhantejani commented on docs/source/quickstart.rst:
```diff
@@ -1,2 +1,157 @@
+
+When an epoch ends we want to run model validation, therefore we attach another handler to the trainer on epoch complete
+event:
+
```
nit: could put a note about passing other args to handlers. Or leave this for another doc/ official docstrings for `add_event_handler`

vfdev-5 commented:
@alykhantejani thanks for the comments! I updated the text according to your comments and added a "(!) Note" as [here](http://pytorch.org/tutorials/beginner/transfer_learning_tutorial.html#convnet-as-fixed-feature-extractor) about "passing other args to handlers"

vfdev-5 commented:
Added basic [concepts](https://github.com/vfdev-5/ignite/blob/issue_108/docs/source/concepts.rst)

alykhantejani commented on docs/source/concepts.rst:
```diff
@@ -1,2 +1,108 @@
+------
+
+The **essence** of the framework is the class :class:`Engine`, an abstraction that loops given number of times over
+provided data, executes a processing and returns a result:
```
processing **function**

alykhantejani commented on docs/source/concepts.rst:
```diff
@@ -1,2 +1,108 @@
+Engine
+------
+
+The **essence** of the framework is the class :class:`Engine`, an abstraction that loops given number of times over
```
loops **a** given

alykhantejani commented:
Thanks @vfdev-5!

vfdev-5 commented:
@alykhantejani do you still want me to integrate your comments ?

alykhantejani commented:
@vfdev-5 if you don't mind, it was very minor though so didn't want to delay merging this. I need to get the docs ready to be hosted and start building the conda tar + wheels for release

vfdev-5 commented:
@alykhantejani no problems, go ahead with your tasks. Thanks!

jasonkriss commented:
@vfdev-5 Thanks a lot! These both look great. I think the only other thing I would maybe add is a mention in the "State" section of "concepts" about the ability to add arbitrary data to the state object in handlers.

vfdev-5 commented:
@jasonkriss yes, this is an important point that completly missed, thanks! At the begining I thought we would fix the scope of the state, especially when talking about engine checkpointing and restoration. As it is more or less fixed, I can add this in another PR.

vfdev-5 commented:
@jasonkriss rethinking about what you said:
>  the ability to add arbitrary data to the state object in handlers.

Do you talk about this (same with handler functions):
```python
import numpy as np
from torch.utils.data import DataLoader
from ignite.engines import Engine

def update(engine, batch):
    engine.state.new_attribute = 12345

engine = Engine(update)

data = np.arange(100)
data = DataLoader(data, batch_size=10)
state = engine.run(data, max_epochs=2)

for attr, value in state.__dict__.items():
    print(attr, value)
```
with the output:
```
iteration 20
epoch 2
max_epochs 2
batch ...

new_attribute 12345

output None
dataloader <torch.utils.data.dataloader.DataLoader object at 0x7f124b81c518>
metrics {}
```
if yes, we can add a new attribute to `state`, `engine` and whatever object...

I thought also to add the method `__repr__` to the class `State` to visualize easily its variable content:
```python
def __repr__(self):
    s = "State:\n"
    for attr, value in self.__dict__.items():
        value = value if not torch.is_tensor(value) else "{} {}".format(type(value), value.shape)
        s += "\t{}: {}\n".format(attr, value)
    return s
```
@alykhantejani @jasonkriss what do you think about ?

jasonkriss commented:
Sorry for the delay on this @vfdev-5 . The `__repr__` sounds like a good idea to me. What do you think @alykhantejani.

@vfdev-5 Did you have another question there? I'm not 100% sure I understood.

vfdev-5 commented:
@jasonkriss no problem, my question was actually about your comment:
> the ability to add arbitrary data to the state object in handlers.

Did you  talk about this ? 
```python
import numpy as np
from torch.utils.data import DataLoader
from ignite.engines import Engine

def update(engine, batch):
    engine.state.new_attribute = 12345

engine = Engine(update)

data = np.arange(100)
data = DataLoader(data, batch_size=10)
state = engine.run(data, max_epochs=2)

for attr, value in state.__dict__.items():
    print(attr, value)
```

jasonkriss commented:
yep exactly. Setting that `new_attribute` is what i was talking about. For me, it usually happens in the handlers but setting it in the `update_fn` is completely valid as well.

vfdev-5 commented:
Right, but actually this is not only ability of `state` but whatever object in python... So, you meant to underline that user can use `state` to add new attributes ?

alykhantejani commented:
@vfdev-5 yes, I think we want to note that this is the preferred way to pass state between handlers. 

When we get to serializing the engine + state, we may have to re-visit this. But for now I think if we establish this user pattern it will be easier

vfdev-5 commented:
Talking about attaching user attributes in `state`, probably, user would like also to initialize his/her attribute before calling it in handler or update function (same as for `metrics` for example). However, today, the state is `None` before running the engine...

jasonkriss commented:
I've been using this feature successfully since we added it. If my attribute is a list or a dict, I will initialize it if its `None` the first time the handler is called. If you want to, you could also have an `initialize` handler that you attach to the start event. Anything more than that would be adding too much complexity IMO. I think the goal of the `state` object is to be a flexible, low-cost abstraction for passing around ad hoc information.

vfdev-5 commented:
Sure, for me `State` is something that describes `engine`, so why state is None, when `engine` is initialized and possibly has handlers etc but not yet run... Maybe, we can initialize it on engine creation and complete with `data`, `metrics` etc on `run()`
